### PR TITLE
Allow for local dns resolution with a custom dialer

### PR DESCRIPTION
**Pull request by stephaniehingtgen**
If the custom dialer updates the transport to proxy the mssql connection, and only the proxy can reach the network that mssql is running in, then any local DNS resolution will fail.

To address this, this PR ensures the custom dialer is used, if defined, to resolve the DNS.

We've been using this logic on [Grafana's forked repo](https://github.com/grafana/go-mssqldb) for the past few months.

### CODE REVIEW

File: protocol.go
Comment (shueybubbles) [inline protocol.go]: >&& c.Dialer != nil { [](http://example.com/codeflow?start=14&length=20)

might there be value in defining an interface to determine if the custom dialer handles DNS lookups? Do we know that all of them do?

Comment (shueybubbles) [inline protocol.go]: >c != nil [](http://example.com/codeflow?start=5&length=8)

I don't think c is ever nil

**Comment by shueybubbles** [comment]
thank you for submitting a PR!

File: protocol.go
Comment (stephaniehingtgen) [reply protocol.go]: It looks like there is a similar check in the [getDialer](https://github.com/microsoft/go-mssqldb/blob/main/mssql.go#L207) method, but I'm happy to remove the check, if that check is there because the method is called in different circumstances

**Comment by stephaniehingtgen** [comment]
@microsoft-github-policy-service agree company="Grafana"

File: protocol.go
Comment (stephaniehingtgen) [reply protocol.go]: My initial thought is that if the custom dialer sends the final request to a different network, that they would want to use the DNS lookup where the actual requests will later be sent to. 

For example, if the driver is being run in a k8s namespace that has a MSSQL instance with the local DNS of `mssql`, but we want to proxy the request to a separate MSSQL instance in another cluster with the local DNS of `mssql2`, we'd want the DNS resolution to happen on the other k8s cluster, otherwise we may get false positives that the DNS is correct if we are using `mssql`, when it won't work (Or vice versa if the DNS is switched):
Screenshot 2023-06-27 at 1 25 46 PM

But if there is a use case for it, I'd be happy to implement this.

Comment (shueybubbles) [reply protocol.go]: I don't think we're in disagreement about the end result. I'm asking if there are multiple types of custom dialers - some that proxy to other networks and some that rely on the current DNS lookup and if we'd need to distinguish between the two. EG, your dialer is clearly of the former. Anyone using the existing model, though, is getting the DNS lookup "for free" and we risk breaking them. 
Maybe a better name for `Dialer` would have been `IPDialer` since the driver has always called it with an IP address up to now. 
Your dialer could be `HostDialer` which takes a host name.
eg
```go
h, ok := c.Dialer.(HostDialer)
```

Comment (stephaniehingtgen) [reply protocol.go]: Sounds good! I'll work on implementing the second `HostDialer` interface

Comment (stephaniehingtgen) [reply protocol.go]: Added :)

**Review by shueybubbles — APPROVED** [review]
Please update the README and the CHANGELOG. Once I get a second approver we will merge.
thx!

**Review by apoorvdeshmukh — CHANGES_REQUESTED** [review]
@stephaniehingtgen Can you look into the [test failure](https://github.com/microsoft/go-mssqldb/actions/runs/5394149307/jobs/9802180755)?

**Comment by stephaniehingtgen** [comment]
@shueybubbles updated the README & Changelog! Thanks for the review!

@apoorvdeshmukh I updated the unit test, but I'm unable to kick off running the tests. Would you be able to kick it off for me?

**Comment by stephaniehingtgen** [comment]
Hey! Just wanted to check if there is anything I need to do here to get this merged. Thanks!

**Comment by shueybubbles** [comment]
@apoorvdeshmukh it's waiting for you to approve then we can merge

**Review by apoorvdeshmukh — APPROVED** [review]
LGTM! Thanks for the fix!

**Pull request merged**
